enterprise/src/main/java/io/crate/enterprise/loader/EnterpriseLoader.java

```diff
@@ -0,0 +1,187 @@
+/*
+ * Licensed to Crate under one or more contributor license agreements.
+ * See the NOTICE file distributed with this work for additional
+ * information regarding copyright ownership.  Crate licenses this file
+ * to you under the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.  You may
+ * obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+ * implied.  See the License for the specific language governing
+ * permissions and limitations under the License.
+ *
+ * However, if you have executed another commercial license agreement
+ * with Crate these terms will supersede the license and you may use the
+ * software solely pursuant to the terms of the relevant commercial
+ * agreement.
+ */
+
+package io.crate.enterprise.loader;
+
+import com.google.common.annotations.VisibleForTesting;
+import io.crate.settings.CrateSetting;
+import io.crate.settings.SharedSettings;
+import org.elasticsearch.common.inject.internal.Nullable;
+import org.elasticsearch.common.settings.Settings;
+
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * Loads classes or values which are only available at runtime.
+ */
+public class EnterpriseLoader {
+
+    private static final Map<Class<?>, Object> loadedInstances = new HashMap<>();
+    private static final Map<Method, Object> loadedValues = new HashMap<>();
```

Comment:
We can't use static maps as cache. In the tests we're running multiple nodes in the same JVM - and these nodes must not share any instances, otherwise our integration tests diverge from a real cluster setup even more.